Add SplitView control

amwx opened the pull request:
## What does the pull request do?
Adds SplitView control from my FluentAvalonia repo with a few enhancements, primarily moving pane animation to Styles

![NewSplitView](https://user-images.githubusercontent.com/40413319/85820297-8cb75500-b73b-11ea-8c5c-0d272dc12ccb.gif)



## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

amwx committed:
Add SplitView

SplitView use nameof

amwx committed:
Adds SplitView styles

amwx committed:
Adds control catalog page

amwx committed:
Add unit tests

amwx commented on src/Avalonia.Themes.Fluent/SplitView.xaml:
```diff
@@ -0,0 +1,221 @@
+  <Style Selector="SplitView:overlay:left /template/ Panel#PART_PaneRoot">
+    <Setter Property="Width" Value="{Binding RelativeSource={RelativeSource TemplatedParent}, Path=TemplateSettings.ClosedPaneWidth}" />
+    <!-- ColumnSpan should be 2 -->
+    <Setter Property="Grid.ColumnSpan" Value="1"/>
```
ColumnSpan should technically be 2 here (and also in :compactoverlay below). When I copied the code into Avalonia from my test project this broke and opening the pane would also move the content slightly, but only when the pane is on the left. Not sure why this is being handled differently...

jmacato replied:
well.. could you try explicitly setting grid.column too? just in case there's some bug in there...

amwx commented on src/Avalonia.Themes.Fluent/SplitView.xaml:
```diff
@@ -0,0 +1,221 @@
+        <Grid Name="Container" Background="{TemplateBinding Background}">
+          <Grid.ColumnDefinitions>
+            <!-- why is this throwing a binding error? -->
+            <ColumnDefinition Width="{Binding RelativeSource={RelativeSource TemplatedParent}, Path=TemplateSettings.PaneColumnGridLength}"/>
```
This is throwing a binding error (you can see it on load or when you change the `PanePlacement`).    
`[Binding] Error in binding to 'Avalonia.Controls.ColumnDefinition'.'Width': 'Null value in expression ''.' (ColumnDefinition #41130254)`
Like my comment below, this only started happening when I moved the code from my test project into avalonia. I know and have checked `TemplateSettings.PaneColumnGridLength` is not null so I'm not sure why this is happening.

jmacato commented:
![Data-yes](https://user-images.githubusercontent.com/16554748/85825829-d04ea500-b7b5-11ea-8c0c-ac8138ed5c03.gif)

amwx committed:
Merge branch 'master' into SplitViewPort

jmacato commented on src/Avalonia.Themes.Fluent/SplitView.xaml:
```diff
@@ -0,0 +1,221 @@
+  </Style>
+
+  <!-- Open/Close Pane animation
+  Uses KeySpline in WinUI, QuinticEaseOut easing seemed to match best
```
keen eye :) i've been meaning to make a SplineEasing or something related so that users can modify that to their liking

jmacato reviewed (approved):
LGTM, just some comments :)

jmacato commented:
@AvaloniaUI/core dont merge this yet.. i have an incoming PR for spline  easing

maxkatz6 commented on src/Avalonia.Controls/SplitView.cs:
```diff
@@ -0,0 +1,476 @@
+    /// <summary>
+    /// Defines constants for how the SplitView Pane should display
+    /// </summary>
+    public enum SplitViewDisplayMode
```
nit: I noticed that many controls is placed in root of Avalonia.Controls folder. Should we start to move controls and related classes/enums to specific sub folders?

danwalmsley replied:
I think that would be a good idea for more complex controls that are not single file... what about namespace too?

maxkatz6 replied:
We should keep Avalonia.Controls, because it is convenient for developers.

amwx replied:
I agree here, I think the controls folder could be better organized this way. Would make it easier to find things too

maxkatz6 commented on src/Avalonia.Controls/SplitView.cs:
```diff
@@ -0,0 +1,476 @@
+            PseudoClasses.Remove($":{oldState}");
+            PseudoClasses.Add($":{newState}");
+
+            var state = (SplitViewDisplayMode)e.NewValue;
+            //PaneColumn Specs: Width/ColumnWidth
+            //   Overlay - 0px/0px
+            //   CompactOverlay - CompactPaneLength/CompactPaneLength
+            //   Inline - 0px/Auto
+            //   CompactInline - CompactPaneLength/Auto
+            TemplateSettings.ClosedPaneWidth = (state == SplitViewDisplayMode.CompactInline ||
+                state == SplitViewDisplayMode.CompactOverlay) ? CompactPaneLength : 0;
+
+            TemplateSettings.PaneColumnGridLength = (state == SplitViewDisplayMode.CompactInline ||
+                state == SplitViewDisplayMode.Inline) ? new GridLength(0, GridUnitType.Auto) :
+                state == SplitViewDisplayMode.Overlay ? new GridLength(0, GridUnitType.Pixel) :
+                new GridLength(CompactPaneLength, GridUnitType.Pixel);
```
I think new switch syntax could fit here.

```suggestion
            var (closedPaneWidth, paneColumnGridLength) = (SplitViewDisplayMode)e.NewValue switch
            {
                SplitViewDisplayMode.Overlay => (0, new GridLength(0, GridUnitType.Pixel)),
                SplitViewDisplayMode.CompactOverlay => (CompactPaneLength, new GridLength(CompactPaneLength, GridUnitType.Pixel)),
                SplitViewDisplayMode.Inline => (0, new GridLength(0, GridUnitType.Auto)),
                SplitViewDisplayMode.CompactInline => (CompactPaneLength, new GridLength(0, GridUnitType.Auto)),
                _ => throw new NotImplementedException(),
            };
            TemplateSettings.ClosedPaneWidth = closedPaneWidth;
            TemplateSettings.PaneColumnGridLength = paneColumnGridLength;
```

amwx committed:
Use new switch syntax

Co-authored-by: Max Katz <maxkatz6@outlook.com>

amwx committed:
Merge branch 'master' into SplitViewPort

Commit:
Merge branch 'master' into SplitViewPort

jmacato committed:
Merge branch 'master' into SplitViewPort

jmacato committed:
Update SplitView.xaml

Update SplitView easings to use what's in UWP's template instead of QuinticEaseOut.

danwalmsley commented on samples/ControlCatalog/ViewModels/SplitViewPageViewModel.cs:
```diff
@@ -0,0 +1,47 @@
+            }
+        }
+
+        private int _displayMode = 3; //CompactOverlay
```
nit: private fields need to be at the top of the class.

danwalmsley commented on src/Avalonia.Controls/SplitView.cs:
```diff
@@ -0,0 +1,473 @@
+        public static readonly StyledProperty<SplitViewTemplateSettings> TemplateSettingsProperty =
+            AvaloniaProperty.Register<SplitView, SplitViewTemplateSettings>(nameof(TemplateSettings));
+
+        private bool _isPaneOpen;
```
these fields should be at top.

amwx replied:
I thought fields went between the AvaloniaProperties and constructors?

danwalmsley replied:
yes your right,

danwalmsley commented on src/Avalonia.Controls/SplitView.cs:
```diff
@@ -0,0 +1,473 @@
+        {
+            base.OnDetachedFromVisualTree(e);
+            if (_pointerDisposables != null)
+                _pointerDisposables.Dispose();
```
`_pointerDisposables?.Dispose();`

danwalmsley commented on src/Avalonia.Controls/SplitView.cs:
```diff
@@ -0,0 +1,473 @@
+        private void OnWindowLostFocus()
+        {
+            if (IsPaneOpen && ShouldClosePane())
+                IsPaneOpen = false;
```
nit: prefer { }

danwalmsley commented on src/Avalonia.Controls/SplitView.cs:
```diff
@@ -0,0 +1,473 @@
+                    break;
+                }
+                src = src.VisualParent;
+            }
```
nit: 
follow } with a newline

danwalmsley commented on src/Avalonia.Controls/SplitView.cs:
```diff
@@ -0,0 +1,473 @@
+        {
+            if (!IsPaneOpen)
+                return;
+            var mouse = obj as RawPointerEventArgs;
```
nit: 
newline after return;

danwalmsley commented on src/Avalonia.Controls/SplitView.cs:
```diff
@@ -0,0 +1,473 @@
+        {
+            var newLen = (double)e.NewValue;
+            var displayMode = DisplayMode;
+            if (displayMode == SplitViewDisplayMode.CompactInline)
```
nit: add a newline before the if

danwalmsley reviewed (commented):
some minor nits.

amwx committed:
Fix formatting

amwx commented:
@danwalmsley  I believe I've addressed all you're comments
@jmacato The new spline easing looks fantastic!

ShadowDancer commented:
@amwx could you explain how exactly overlay mode works? When I try render content over other controls they appear on top.

amwx commented:
@ShadowDancer In Overlay mode, the pane is supposed to display above content. Tho there is currently an issue with the SplitView in Overlay mode, because of a difference in behavior between the Avalonia source and the nuget package for some reason. So currently, Overlay mode displays like Inline.
If you can open up the ControlCatalog, the example in there works as intended for all four display modes

ShadowDancer commented:
@amwx how exactly you make pane display above the content? I do not understand how this works, when I try replicating it content is displayed on top of each other.
